subsys/bluetooth/mesh/pb_adv.c

```diff
@@ -132,28 +149,125 @@ static void link_close(struct prov_rx *rx, struct net_buf_simple *buf);
 static void prov_link_close(enum prov_bearer_link_status status);
 static void close_link(enum prov_bearer_link_status status);
 
-static void buf_sent(int err, void *user_data)
+static void tx_work_handler(struct k_work *work);
+static K_WORK_DELAYABLE_DEFINE(tx_work, tx_work_handler);
+
+static void tx_schedule(void)
 {
-	enum prov_bearer_link_status reason = (enum prov_bearer_link_status)(int)user_data;
+	uint16_t random_delay;
+
+	if (atomic_test_bit(link.flags, ADV_SENDING)) {
+		LOG_DBG("Another tx is in progress");
+		return;
+	}
+
+	(void)bt_rand(&random_delay, sizeof(random_delay));
+	random_delay = 20 + (random_delay % 30);
+
+	LOG_DBG("Next PDU delayed by %ums", random_delay);
+
+	(void)k_work_schedule(&tx_work, K_MSEC(random_delay));
+}
+
+static void send_unacked(struct bt_mesh_adv *adv, prov_bearer_send_complete_t cb,
+			 void *cb_data)
+{
+	for (int i = 0; i < ARRAY_SIZE(link.tx.unacked); i++) {
+		if (link.tx.unacked[i].adv != NULL) {
+			continue;
+		}
+
+		link.tx.unacked[i].adv = adv;
+		link.tx.unacked[i].cb = cb;
+		link.tx.unacked[i].cb_data = cb_data;
+
+		tx_schedule();
 
-	if (atomic_test_and_clear_bit(link.flags, ADV_LINK_CLOSING)) {
-		close_link(reason);
 		return;
 	}
+
+	LOG_WRN("No memory to send unacked PDU: %s", bt_hex(adv->b.data, adv->b.len));
```

Comment:
Is it possible at all?

Reply:
I may guess that in theory possible if number of retransmissions for acks is high enough so that we are still sending Link Ack and Gen Trans Ack from the previous transaction and receive a new transaction, or sending Gen Trans Ack from the previous transaction and a new one and trying to send Link Close. But the protocol will retry again if new ack is missing, so this unlikely will cause a problem. I added return code here to handle this when trying to send Gen Trans Ack or Link Close.